File: Python/addStrings.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: addStrings.py
@Function: 字符串相加 数学问题
@Link: https://leetcode-cn.com/problems/add-strings/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-18
"""


class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        res = ""
        i, j, carry = len(num1) - 1, len(num2) - 1, 0
        while i >= 0 or j >= 0:
            n1 = int(num1[i]) if i >= 0 else 0
            n2 = int(num2[j]) if j >= 0 else 0
            tmp = n1 + n2 + carry
            carry = tmp // 10
            res = str(tmp % 10) + res
            i, j = i - 1, j - 1
        return "1" + res if carry else res


# ----------------------------------
if __name__ == "__main__":
    num1 = "99"
    num2 = "1"

    solution = Solution()
    add_sum = solution.addStrings(num1, num2)
    print(f"The solution of this problem is : {add_sum}")

File: C++/addStrings.cpp
/**
 * @File    : addStrings.cpp
 * @Brief   : 字符串相加 数学问题
 * @Link    : https://leetcode-cn.com/problems/add-strings/
 * @Author  : Wei Li
 * @Date    : 2021-07-18
*/

#include <iostream>
#include <string>
#include <algorithm>

class Solution
{
public:
    std::string addStrings(std::string num1, std::string num2)
    {
        std::string output("");
        std::reverse(num1.begin(), num1.end());
        std::reverse(num2.begin(), num2.end());
        int onelen = num1.length(), twolen = num2.length();
        if (onelen <= twolen)
        {
            std::swap(num1, num2);
            std::swap(onelen, twolen);
        }
        int addbit = 0;
        for (int i = 0; i < twolen; ++i)
        {
            int cur_sum = (num1[i] - '0') + (num2[i] - '0') + addbit;
            output += std::to_string((cur_sum) % 10);
            addbit = cur_sum < 10 ? 0 : 1;
        }
        for (int i = twolen; i < onelen; ++i)
        {
            int cur_sum = (num1[i] - '0') + addbit;
            output += std::to_string((cur_sum) % 10);
            addbit = cur_sum < 10 ? 0 : 1;
        }
        if (addbit)
        {
            output += "1";
        }
        std::reverse(output.begin(), output.end());
        return output;
    }
};

// -----------------------------
int main(int argc, char **argv)
{
    std::string num1 = "99", num2 = "1";

    Solution solution;
    auto sum_str = solution.addStrings(num1, num2);
    std::cout << "The solution of this problem is " << sum_str << std::endl;

    return 0;
}
